lib/mesh.h

```diff
@@ -0,0 +1,337 @@
+#pragma once
+
+#include "functional.h"
+#include "triangle.h"
+#include "types.h"
+
+#include <OpenMesh/Core/Mesh/PolyMesh_ArrayKernelT.hh>
+#include <OpenMesh/Core/Utils/PropertyManager.hh>
+
+#include <array>
+#include <unordered_map>
+
+using RadiosityMesh = OpenMesh::PolyMesh_ArrayKernelT<>;
+
+// Property to store the vertices describing corners of a triangle
+// Since after subdivision, a triangle may consists of more than 3 vertices, we
+// have to keep track of the corners.
+using CornerVertices =
+    OpenMesh::FPropHandleT<std::array<RadiosityMesh::VertexHandle, 3>>;
+using CornerVerticesProperty =
+    OpenMesh::PropertyManager<CornerVertices, RadiosityMesh>;
+
+using VertexRadiosityHandle = OpenMesh::VPropHandleT<Color>;
+using VertexRadiosityHandleProperty =
+    OpenMesh::PropertyManager<VertexRadiosityHandle, RadiosityMesh>;
+
+using FaceRadiosityHandle = OpenMesh::FPropHandleT<Color>;
+using FaceRadiosityHandleProperty =
+    OpenMesh::PropertyManager<FaceRadiosityHandle, RadiosityMesh>;
+
+namespace detail {
+auto get_halfedge_handle(const RadiosityMesh& mesh,
+                         RadiosityMesh::FaceHandle face, size_t index) {
+    auto it = mesh.cfh_iter(face);
+    while (index--) {
+        it++;
+    }
+    return *it;
+}
+
+auto find_vertex(const RadiosityMesh& mesh, RadiosityMesh::FaceHandle face,
+                 const RadiosityMesh::Point& pt) {
+    size_t index = 0;
+    for (auto v : mesh.fv_range(face)) {
+        if (mesh.point(v) == pt) {
+            break;
+        }
+        index++;
+    }
+    return index;
+}
+} // namespace detail
+
+auto build_mesh(const Triangles& triangles) {
+    RadiosityMesh mesh;
+    // a vertex is uniquely determined by its coordinates and a normal
+    std::unordered_map<std::pair<Vec, Vec>, RadiosityMesh::VertexHandle> lookup;
+
+    auto corners_prop =
+        CornerVerticesProperty::createIfNotExists(mesh, "corner_vertices");
+
+    std::array<RadiosityMesh::VertexHandle, 3> vhandles;
+    for (const auto& tri : triangles) {
+        const Vec& a = tri.vertices[0];
+        const Vec& b = tri.vertices[1];
+        const Vec& c = tri.vertices[2];
+        const Vec& na = tri.normals[0];
+        const Vec& nb = tri.normals[1];
+        const Vec& nc = tri.normals[2];
+
+        auto ita = lookup.emplace(std::make_pair(a, na),
+                                  RadiosityMesh::VertexHandle());
+        if (ita.second) {
+            ita.first->second = mesh.add_vertex({a.x, a.y, a.z});
+        }
+        auto va = ita.first->second;
```

Comment:
Nein, der Typ ist `RadiosityMesh::VertexHandle`. C++ ist leider in der Hinsicht echt nicht gut. Man braucht einfach nur named Tuples.